### PR TITLE
[Loki log handler] - Integration with Grafana Loki

### PR DESCRIPTION
**Pull request by whatnick**
This Pull request will make the following changes:

Create a log handler that pushes task logs to Loki with labels and using the same labels pull logs to Airflow Web UI. This is a bump of the previous PR by @sshah90 and attempts to cherry pick changes for review and integration of maintainer feedback.

related: 
#12920
#12925 

Note:

I had added documentation at [logging-tasks.rst](https://github.com/apache/airflow/blob/master/docs/apache-airflow/logging-monitoring/logging-tasks.rst) file but looks like all log handler docs have been moved to providers folders so I make change accordingly and update the same PR.

### CODE REVIEW

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        :param task_instance: task instance object
+        :type: task_instance: TaskInstance
+        """
+        # Not adding execution date since it violates Loki label standards
+        # https://grafana.com/blog/2020/08/27/the-concise-guide-to-labels-in-loki/
+
+        return {
+            "airflow_dag_id": task_instance.dag_id,
+            "airflow_task_id": task_instance.task_id,
+            "airflow_try_number": str(task_instance.try_number),
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:108]: Should run_id be included here? For dags where multiple runs go in parallel. 
Note though, that dynamic labels such as run_id are discouraged according to https://grafana.com/docs/loki/latest/best-practices/ 
It's a little odd that try_number is included and run_id is not. If you want to flatten it into single stream i would argue try_number should not be included either.

We use Loki for airflow tasks today (but ship the logs via Promtail) What we have done is to use dag_id and task_id as labels, then "[pack](https://grafana.com/docs/loki/latest/clients/promtail/stages/pack/)" the run_id into the stream. Since the task start+end date limits the stream length it becomes a short search.  Anecdotally It's working fine for us, but our volumes are not very big either.

Comment (whatnick) [reply airflow/providers/grafana/log/loki_task_handler.py:108]: Thanks for the review, I am redoing this PR from an old PR which was abandoned. I am aware of the curse of dimensionality the addition of dynamic labels creates. Better to search by those than index and sort by those. Could you summarize in tabular form the metadata in labels and those as contents of the stream to improve this PR.

Also would love to see a diagram of your work around via promtail to loki to share with the team. FYI I started on this task as part of my work towards [DEA Airflow](https://github.com/GeoscienceAustralia/dea-airflow).

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    def _read(self, task_instance, try_number, metadata=None) -> Tuple[str, Dict[str, bool]]:
+        """
+        Return list of logs string for each task try number by streaming it from Loki
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        :param try_number: current try_number to read log from
+        :type: try_number: int
+        """
+        logs = ""
+        if self.is_loki_alive() == 200:
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:221]: IMO this type of pre-is-alive-check is redundant, better just try the real query right away and only fallback on any exceptions.

Checking for is_alive before:
* Introduces unnecessary delay
* Introduces potential race condition where alive at first works, but 1ms later server goes down and query fails. Forcing you to deal with the query-failure scenario anyway.
* Introduces one more source of error, in case the is_alive fails because of intermittent network flakyness.

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        start_epoch_time = self.convert_epoch_time(start_time)
+
+        if end_time and next_try < 3:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            return (start_epoch_time - 60, end_epoch_time + 60 * 2)
+        if end_time:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            execution_time_epoch_time = self.convert_epoch_time(execution_time)
+            return (execution_time_epoch_time - 60, end_epoch_time + 60 * 2)
+        return (start_epoch_time - 60, start_epoch_time + 60 * 60 * 24)
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:183]: Should end of range here be datetime.now instead? Otherwise it will not be possible to see logs for tasks in progress that have been running for over 24 hours. Maybe use `task_instance.execution_timeout` if set ?

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                f"and end_time {query_parameters['end']}\n"
+            )
+            try:
+                response = self.session.get(
+                    f"{self.get_conn.host}/loki/api/v1/query_range",
+                    params=query_parameters,
+                ).json()
+                results = response["data"]["result"]
+                logs_list = [value for result in results for value in result["values"]]
+                # sorting with list of values with timestamp.
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:235]: Why is this sorting needed? Is it in case multiple streams are returned? Because within one stream the result is already sorted when returned from loki.

https://grafana.com/docs/loki/latest/api/#query-loki-over-a-range-of-time

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                logs += "\n".join([f"{log_line[1]}" for log_line in sorted(logs_list, key=lambda x: x[0])])
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:236]: Maybe append some kind of warning message to the log if the max number of values per query is reached. 

https://grafana.com/docs/loki/latest/api/#query-loki-over-a-range-of-time

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        start_time, end_time, execution_time = (
+            str(task_instance.start_date.isoformat()),
+            task_instance.end_date,
+            str(task_instance.execution_date.isoformat()),
+        )
+        next_try = task_instance.next_try_number
+        start_epoch_time = self.convert_epoch_time(start_time)
+
+        if end_time and next_try < 3:
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:176]: What does `< 3 `mean here?

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        start_time, end_time = self.query_time_range(task_instance)
+        task_lables = self.get_label(task_instance)
+        del task_lables["airflow_try_number"]  # try_number will come from read method
+
+        # converting task_lables(dict) to LogQL format
+        # {"dag_id":"airflow"} --> {dag_id = "airflow"}
+        query = "{"
+        for key, value in task_lables.items():
+            query += f"""{key}="{value}", """
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:201]: Does this need any escaping or sanitizing to avoid query injections?

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        query += f"""airflow_try_number="{try_number}"}} != "WARNING" """
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:202]: Why `!= "WARNING"` ?

File: airflow/providers/grafana/log/loki_task_handler.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+
+class LokiTaskHandler(FileTaskHandler, LoggingMixin):
+    """
+    LokiTaskHandler that directly makes Loki logging API calls while reading and writing logs.
+    This is a Python standard ``logging`` handler using that can be used to route Python standard
+    logging messages directly to the Loki Logging API. It can also be used to save logs for
+    executing tasks. To do this, you should set as a handler with the name "tasks". In this case,
+    it will also be used to read the log for display in Web UI.
+    :param base_log_folder: Base log folder to place logs (incase Loki is down).
```
Comment (hterik) [inline airflow/providers/grafana/log/loki_task_handler.py:42]: (I would advice you to always place logs locally as files, and use a daemon like Promtail to ship them to remote storage. 
This decouples the logging from the application and improves reliability for both of them. )

**Review by hterik — COMMENTED** [review]
I'm not really guardian of this project so take my comments as recommendations only. Just possibly interested in this module myself in future.

File: airflow/providers/grafana/log/loki_task_handler.py (hunk 4 above)
Comment (hterik) [reply airflow/providers/grafana/log/loki_task_handler.py:235]: In case of multiple streams, i believe this sorting can be done more efficiently as well, since each individual stream is already sorted. 
Instead of putting all values together into one logs_list and then sorting the whole list, you can iterate one item at a time from each of the value-lists, but only pick the one with highest timestamp when appending to the final log list. This should be O(N) rather than O(N logN) and one less list-comprehension-in-list-comprehension.

**Comment by snjypl** [comment]
i would like to mention a grafana loki  provider that i am working on. 

https://github.com/snjypl/airflow-provider-grafana-loki

everyone's feedback would be greatly appreciated.

**Comment by potiuk** [comment]
@whatnick - are you sure you want to continue on that and try to make contribution to the apache project if there is another provider someone maintains? There are a number of conditions that have to be fulfilled to add a provider to the Airlfow repo (it has to be really maintaineable long-term by the community) - and if somoene wants to maintain an integration outside (and for example puts a link on the ecosystem page https://airflow.apache.org/ecosystem/ ), that's quite a bit less for us to maintain, and easier to relaease and produce such 3rd-party providers. Shall we close this PR?
